numpy/core/src/common/npy_dlpack.h

```diff
@@ -0,0 +1,30 @@
+#include "Python.h"
+#include "dlpack/dlpack.h"
+
+#ifndef NPY_DLPACK_H
+#define NPY_DLPACK_H
+
+// Part of the Array API specification.
+#define NPY_DLPACK_CAPSULE_NAME "dltensor"
+#define NPY_DLPACK_USED_CAPSULE_NAME "used_dltensor"
+
+// Used internally by NumPy to store a base object
+// as it has to release a reference to the original
+// capsule.
+#define NPY_DLPACK_INTERNAL_CAPSULE_NAME "numpy_dltensor"
+
+static void array_dlpack_capsule_deleter(PyObject *self)
+{
+    if (!PyCapsule_IsValid(self, NPY_DLPACK_CAPSULE_NAME) &&
+            !PyCapsule_IsValid(self, NPY_DLPACK_INTERNAL_CAPSULE_NAME)) {
+        if (!PyCapsule_IsValid(self, NPY_DLPACK_USED_CAPSULE_NAME)) {
+            PyErr_SetString(PyExc_RuntimeError, "Invalid capsule name.");
+        }
+        return;
+    }
+    DLManagedTensor *managed = 
+        (DLManagedTensor *)PyCapsule_GetPointer(self, PyCapsule_GetName(self));
+    managed->deleter(managed);
+}
```

Comment:
If we split this in two, then we can be a little clear about what's dlpack boilerplate and what's our design:
```suggestion
/* This is exactly as mandated by dlpack (and if it isn't, it should be!) */
static void array_dlpack_capsule_deleter(PyObject *self)
{
    if (PyCapsule_IsValid(self, NPY_DLPACK_USED_CAPSULE_NAME)) {
        return;
    }
    DLManagedTensor *managed = 
        (DLManagedTensor *)PyCapsule_GetPointer(self, NPY_DLPACK_CAPSULE_NAME);
    if (managed == NULL) {
        return NULL;
    }
    managed->deleter(managed);
}

/* used internally */
static void array_dlpack_internal_capsule_deleter(PyObject *self)
{
    DLManagedTensor *managed = 
        (DLManagedTensor *)PyCapsule_GetPointer(self, NPY_DLPACK_INTERNAL_CAPSULE_NAME);
    if (managed == NULL) {
        return NULL;
    }
    managed->deleter(managed);
}
```

Reply:
Sigh, this change made the warnings come back, even though both deleters are being used in different parts of the codebase.

Reply:
Maybe just move the second one to where it's actually used?